grr/server/grr_response_server/databases/spanner_flows.py

```diff
@@ -0,0 +1,2512 @@
+        if msgs:
+          for m in msgs:
+            self.flow_processing_request_handler_pool.AddTask(
+                target=handler, args=(m,)
+            )
+        else:
+          time.sleep(self._FLOW_REQUEST_POLL_TIME_SECS)
+
+      except Exception as e:  # pylint: disable=broad-except
+        logging.exception("_FlowProcessingRequestHandlerLoop raised %s.", e)
+        time.sleep(self._FLOW_REQUEST_POLL_TIME_SECS)
+
+    self.flow_processing_request_handler_pool.Stop()
+
+  def RegisterFlowProcessingHandler(
+      self, handler: Callable[[flows_pb2.FlowProcessingRequest], None]
+  ):
+    """Registers a handler to receive flow processing messages."""
+    self.UnregisterFlowProcessingHandler()
+
+    if handler:
+      self.flow_processing_request_handler_stop = False
+      self.flow_processing_request_handler_thread = threading.Thread(
+          name="flow_processing_request_handler",
+          target=self._FlowProcessingRequestHandlerLoop,
+          args=(handler,),
+      )
+      self.flow_processing_request_handler_thread.daemon = True
+      self.flow_processing_request_handler_thread.start()
+
+  def UnregisterFlowProcessingHandler(
+      self, timeout: Optional[rdfvalue.Duration] = None
+  ) -> None:
+    """Unregisters any registered flow processing handler."""
+    if self.flow_processing_request_handler_thread:
+      self.flow_processing_request_handler_stop = True
+      self.flow_processing_request_handler_thread.join(timeout)
+      if self.flow_processing_request_handler_thread.is_alive():
+        raise RuntimeError("Flow processing handler did not join in time.")
+      self.flow_processing_request_handler_thread = None
+
+  @db_utils.CallLogged
+  @db_utils.CallAccounted
+  def WriteFlowRequests(
+      self,
+      requests: Collection[flows_pb2.FlowRequest],
+  ) -> None:
+    """Writes a list of flow requests to the database."""
+
+    flow_keys = [(r.client_id, r.flow_id) for r in requests]
+
+    def Txn(txn) -> None:
+      needs_processing = {}
+      columns = ["ClientId",
+                 "FlowId",
+                 "RequestId",
+                 "NeedsProcessing",
+                 "NextResponseId",
+                 "CallbackState",
+                 "Payload",
+                 "CreationTime",
+                 "StartTime"]
+      rows = []
+      for r in requests:
+        if r.needs_processing:
+          needs_processing.setdefault((r.client_id, r.flow_id), []).append(r)
+
+        if r.start_time:
+          start_time = rdfvalue.RDFDatetime.FromMicrosecondsSinceEpoch(r.start_time).AsDatetime()
+        else:
+          start_time = rdfvalue.RDFDatetime.FromMicrosecondsSinceEpoch(0).AsDatetime()
+
+        rows.append([r.client_id, r.flow_id, str(r.request_id), r.needs_processing, str(r.next_response_id),
+                     r.callback_state, r, spanner_lib.COMMIT_TIMESTAMP, start_time])
+        txn.insert_or_update(table="FlowRequests", columns=columns, values=rows)
```

Comment:
```suggestion
      for r in requests:
        if r.needs_processing:
          needs_processing.setdefault((r.client_id, r.flow_id), []).append(r)

        if r.start_time:
          start_time = rdfvalue.RDFDatetime.FromMicrosecondsSinceEpoch(r.start_time).AsDatetime()
        else:
          start_time = rdfvalue.RDFDatetime.FromMicrosecondsSinceEpoch(0).AsDatetime()

        rows.append([r.client_id, r.flow_id, str(r.request_id), r.needs_processing, str(r.next_response_id),
                     r.callback_state, r, spanner_lib.COMMIT_TIMESTAMP, start_time])
      txn.insert_or_update(table="FlowRequests", columns=columns, values=rows)
```
Let's must the transaction outside of this request loop.